proxy/src/redis/keys.rs

```diff
@@ -0,0 +1,92 @@
+use anyhow::Ok;
+use pq_proto::{id_to_cancel_key, CancelKeyData};
+use serde::{Deserialize, Serialize};
+use std::io::ErrorKind;
+
+pub mod keyspace {
+    pub const CANCEL_PREFIX: &str = "cancel";
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub(crate) enum KeyPrefix {
+    #[serde(untagged)]
+    Cancel(CancelKeyData),
+}
+
+impl KeyPrefix {
+    pub(crate) fn build_redis_key(&self) -> anyhow::Result<String> {
+        match self {
+            KeyPrefix::Cancel(key) => {
+                let key_str = format!("{key}");
+                if let Some(key_str) = key_str
+                    .strip_prefix("CancelKeyData(")
+                    .and_then(|s| s.strip_suffix(')'))
+                {
+                    Ok(format!("{}:{}", keyspace::CANCEL_PREFIX, key_str))
+                } else {
+                    Err(anyhow::anyhow!("Invalid key format"))
+                }
+            }
+        }
+    }
```

Comment:
Maybe better instead to do
```suggestion
    pub(crate) fn build_redis_key(&self) -> String {
        match self {
            KeyPrefix::Cancel(key) => {
                let hi = (key.backend_pid as u64) << 32;
                let lo = (key.cancel_key as u64) & 0xffffffff;
                let id = hi | lo;
                format!("{keyspace::CANCEL_PREFIX}:{id:x}")
            }
        }
    }
```

(extracting the format code directly from `CancelKeyData::fmt`)